README.md
# trimble_gnss_driver
A ROS driver for Trimble GNSS GSOF messages


CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package trimble_gnss_driver
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

0.0.3 (2022-07-07)
------------------
* Merge branch 'master' of https://github.com/SAGARobotics/trimble_gnss_driver
* Position covariance reorder to be ENU
* Contributors: MikHut

0.0.2 (2022-07-07)
------------------
* get position type from receiver msg 38
* Contributors: MikHut

0.0.1 (2022-07-04)
------------------
* remove unused deps
* remove unused deps
* Merge pull request `#3 <https://github.com/MikHut/trimble_gnss_driver/issues/3>`_ from SAGARobotics/Add-launch-file-to-cmakelists
  Update CMakeLists.txt
* Update CMakeLists.txt
* Create internal-release.yml (`#2 <https://github.com/MikHut/trimble_gnss_driver/issues/2>`_)
  * Create internal-release.yml
* Update ci.yml
* Update ci.yml
* Create ci.yml
* Merge pull request `#1 <https://github.com/MikHut/trimble_gnss_driver/issues/1>`_ from MikHut/base_name
  Base name
* python version independant
* checksum fixed
* type 35 msg working, base info, does not work for rtcm though
* initial test for getting the base station name currently used
* GPL license added
* Update package.xml
* tf exec depend
* print heading offset applied to yaw output
* tidy prints, just print warning when msgs are being skipped from being published due to no matching error or invalid data
* option to specifically name the output frame_id and ability to offset the heading using the robot urdf if desired
* dont publish if checksum is false
* move gps qualities dict out of main code
* cleanup
* update comment in launch file
* send attitude as imu msg type instead
* tidying, dont publish zeros, publish yaw on an Imu msg if from dual antenna with no fusion
* added publishers, made nice parse, tidied, sorted pitch coord system for ros
* updates to test on robot
* small fix
* updates
* work in progress, initial setup to test on robot
* Initial commit
* Contributors: Bård-Kristian, Michael Hutchinson, MikHut


src/trimble_gnss_driver/parser.py
#!/usr/bin/env python
"""
GSOF parser

@author: Michael Hutchinson (mhutchinson@sagarobotics.com)

"""


parse_maps = {
    1: [
        ('GPS_TIME',
         'GPS_WEEK',
         'SVN_NUM',
         'FLAG_1',
         'FLAG_2',
         'INIT_NUM'),
        ('>LH4B')
       ],
    2: [
        ('LATITUDE',
         'LONGITUDE',
         'HEIGHT_WGS84'),
        ('>3d'),
       ],
    3: [
        ('X_POS',
         'Y_POS',
         'Z_POS'),
        ('>3d')
       ],
    4: [
        ('LOCAL_DATUM_ID',
         'LOCAL_DATUM_LAT',
         'LOCAL_DATUM_LON',
         'LOCAL_DATUM_HEIGHT',
         'OPRT'),
        ('>8s3dB')
       ],
    5: [
        ('LOCAL_DATUM_ID',
         'LOCAL_ZONE_ID',
         'LOCAL_ZONE_NORTH',
         'LOCAL_ZONE_EAST',
         'LOCAL_DATUM_HEIGHT'),
        ('>2s3d')
       ],
    6: [
        ('DELTA_X',
         'DELTA_Y',
         'DELTA_Z'),
        ('>3d')
       ],
    7: [
        ('DELTA_EAST',
         'DELTA_NORTH',
         'DELTA_UP'),
        ('>3d')
       ],
    8: [
        ('VEL_FLAG',
         'VELOCITY',
         'HEADING',
         'VERT_VELOCITY'),
        ('>B3f')
       ],
    9: [
        ('PDOP',
         'HDOP',
         'VDOP',
         'TDOP'),
        ('>4f')
       ],
    10:[
        ('CLOCK_FLAG',
         'CLOCK_OFFSET',
         'FREQ_OFFSET'),
        ('>B2d')
       ],
    11:[
        ('POSITION_RMS_VCV',
         'VCV_XX',
         'VCV_XY',
         'VCV_XZ',
         'VCV_YY',
         'VCV_YZ',
         'VCV_ZZ',
         'UNIT_VAR_VCV',
         'NUM_EPOCHS_VCV'),
        ('>8fh')
       ],
    12:[
        ('POSITION_RMS_SIG',
         'SIG_EAST',
         'SIG_NORT',
         'COVAR_EN',
         'SIG_UP',
         'SEMI_MAJOR',
         'SEMI_MINOR',
         'ORIENTATION',
         'UNIT_VAR_SIG',
         'NUM_EPOCHS_SIG'),
        ('>9fh')
       ],
    15:[
        ('SERIAL_NUM'),
        ('>l')
       ],
    16:[
        ('GPS_MS_OF_WEEK',
         'CT_GPS_WEEK',
         'UTC_OFFSET',
         'CT_FLAGS'),
        ('>l2hB')
       ],
    26:[
        ('UTC_MS_OF_WEEK',
         'UTC_GPS_WEEK',
         'UTC_SVS_NUM',
         'UTC_FLAG_1',
         'UTC_FLAG_2',
         'UTC_INIT_NUM'),
        ('>lh4B')
       ],
    27:[
        ('GPS_TIME',
        'FLAGS',
        'SVN_NUM',
        'CALC_MODE',
        'RESERVED',
        'PITCH',
        'YAW',
        'ROLL',
        'MASTER_SLAVE_RANGE',
        'PDOP',
        'PITCH_VAR',
        'YAW_VAR',
        'ROLL_VAR',
        'PITCH_YAW_COVAR',
        'PITCH_ROLL_COVAR',
        'YAW_ROLL_COVAR',
        'MASTER_SLAVE_RANGE_VAR'),
        ('>L4B4dH7f')
       ],
    # 34 needs to be dealt with carefully/differently as it depends on the number of
    # satellites in view.
    # 34:[
    #     ('PRN',
    #      'SV_SYSTEM',
    #      'SV_FLAG1',
    #      'SV_FLAG2',
    #      'ELEVATION',
    #      'AZIMUTH',
    #      'SNR_L1',
    #      'SNR_L2',
    #      'SNR_L5'),
    #     ('>5Bh3B')
    #    ],
    35:[
        ('FLAGS',
         'BASE_NAME',
         'BASE_ID_1',
         'BASE_ID_2',
         'BASE_LATITUDE',
         'BASE_LONGITUDE',
         'BASE_HEIGHT'),
        ('>B8s2B3d')
       ],
    37:[
        ('BATT_CAPACITY',
         'REMAINING_MEM'),
        ('>hd')
       ],
    41:[
        ('GPS_TIME_REF',
         'GPS_WEEK_REF',
         'LAT_REF',
         'LONG_REF',
         'HEIGHT_REF',
         'QI'),
        ('>LH3dB')
       ],
    49:[
        ('GPS_WEEK',
         'GPS_TIME',
         'IMU_ALIGNMENT_STATUS',
         'GPS_QUALITY',
         'FUSED_LATITUDE',
         'FUSED_LONGITUDE',
         'FUSED_ALTITUDE',
         'FUSED_VEL_NORTH',
         'FUSED_VEL_EAST',
         'FUSED_VEL_DOWN',
         'FUSED_TOTAL_SPEED',
         'FUSED_ROLL',
         'FUSED_PITCH',
         'FUSED_YAW',
         'FUSED_TRACK_ANGLE',
         'FUSED_VEL_ROLL',
         'FUSED_VEL_PITCH',
         'FUSED_VEL_YAW',
         'FUSED_ACC_X',
         'FUSED_ACC_Y',
         'FUSED_ACC_Z'),
        ('>HL2B3d4f4d6f')
       ],
    50:[
        ('GPS_WEEK',
         'GPS_TIME',
         'IMU_ALIGNMENT_STATUS',
         'GPS_QUALITY',
         'FUSED_RMS_LATITUDE',
         'FUSED_RMS_LONGITUDE',
         'FUSED_RMS_ALTITUDE',
         'FUSED_RMS_VEL_NORTH',
         'FUSED_RMS_VEL_EAST',
         'FUSED_RMS_VEL_DOWN',
         'FUSED_RMS_ROLL',
         'FUSED_RMS_PITCH',
         'FUSED_RMS_YAW'),
        ('>HL2B9f')
       ]
    }


src/trimble_gnss_driver/conversions.py
#!/usr/bin/env python
"""
Utilities for various geodetic conversions. Adapted from https://kb.unavco.org/kb/article/trimble-netr9-receiver-gsof-messages-806.html
"""


from math import sin, cos, sqrt, atan2, degrees, radians


def llh2xyz(lat, lon, h, a=6378137.0, finv=298.257223563, input_degrees=True):
    """
    Converts ellipsoidal coordinates to cartesian.

    :param lat:
    :param lon:
    :param h:

    :return: A tuple containing cartesian coordinates in ECEF [X, Y, Z] meters.
    """
    # convert degrees to radians if necessary
    if input_degrees:
        lat = radians(lat)
        lon = radians(lon)
    # Default (a=6378137.0, finv=298.257223563) is for WGS84
    f = 1/finv
    b = a*(1 - f)
    e2 = 1 - (1 - f)**2
    # Compute the Cartesian coordinates
    v = a/sqrt(1-e2*sin(lat)*sin(lat))
    x = (v+h)*cos(lat)*cos(lon)
    y = (v+h)*cos(lat)*sin(lon)
    z = (v*(1-e2)+h)*sin(lat)
    return x, y, z

def xyz2llh(x, y, z, a=6378137.0, finv=298.257223563, output_degrees=True):
    """
    Converts cartesian coordinates to ellipsoidal. Uses an iterative algorithm.

    :param x: x coordinate in meters
    :param y: y coordinate in meters
    :param z: z coordinate in meters
    :param a: major semi-axis (Default: WGS84)
    :param finv: inverse of flattening (Default: WGS84)
    :param output_degrees: A boolean to select output in degrees (default is True)
    :return: A tuple containing geodetic coordinates in WGS84 (lat, lon, h).

    WGS84: a=6378137.0, finv=298.257223563
    """
    # Default (a=6378137.0, finv=298.257223563) is for WGS84
    f = 1/finv
    b = a*(1 - f)
    e2 = 1 - (1 - f)**2
    # Latitude and height convergence criteria
    elat = 1.0e-12
    eht = 1.e-5
    # Initial values for iteration
    p = sqrt(x*x + y*y)
    lat = atan2(z, p*(1-e2))
    h = 0
    dh = 1
    dlat = 1
    # Iterate until lat & h converge to elat & eht
    while dlat > elat or dh > eht:
        lat0 = lat
        h0 = h
        v = a/sqrt(1-e2*sin(lat)*sin(lat))
        h = p/cos(lat)-v
        lat = atan2(z, p*(1-e2*v/(v+h)))
        dlat = abs(lat-lat0)
        dh = abs(h-h0)
    lon = atan2(y, x)
    if output_degrees:
        # output degrees
        geodetic_coordinates = (degrees(lat), degrees(lon), h)
    else:
        # output radians
        geodetic_coordinates = (lat, lon, h)
    return geodetic_coordinates


def xyz2neu(X, Y, Z, x, y, z):
    """
    Rotates the <(x-X),(y-Y),(z-Z)> vector into a topocentric coordinate system
    tangential to the reference point XYZ.

    Assumes coordinates are given in meters.
    :param X: Reference Coordinate X (meters)
    :param Y: Reference Coordinate Y (meters)
    :param Z: Reference Coordinate Z (meters)
    :param x: Observed location x (meters)
    :param y: Observed location y (meters)
    :param z: Observed location z (meters)
    :return: A tuple containing topocentric coordinates in n, e, u (meters)

    WGS84: a=6378137.0, finv=298.257223563
    """
    lat, lon, h = xyz2llh(X, Y, Z, a=6378137.0, finv=298.257223563, output_degrees=False)
    e = -sin(lon) * (x - X) + cos(lon) * (y - Y)
    n = -sin(lat) * cos(lon) * (x - X) - sin(lat) * sin(lon) * (y - Y) + cos(lat) * (z - Z)
    u = cos(lat) * cos(lon) * (x - X) + cos(lat) * sin(lon) * (y - Y) + sin(lat) * (z - Z)
    return n, e, u


def xyz_to_globk_neu(X, Y, Z, x, y, z):
    """
    Finds the arc-length between two LLH points in dN and dE.
    The parameter dU is defined as the difference between the observed and reference ellipsoidal height.

    Assumes coordinates are given in meters.
    :param X: Reference Coordinate X (meters)
    :param Y: Reference Coordinate Y (meters)
    :param Z: Reference Coordinate Z (meters)
    :param x: Observed location x (meters)
    :param y: Observed location y (meters)
    :param z: Observed location z (meters)
    :return: A tuple containing n, e, u (meters)
    """

    def sub(a, b):
        """
        Subtracts elements in two lists
        """
        c = [i - j for i, j in zip(a, b)]
        return c

    def mul(a, b):
        """
        Multiplies elements in two lists
        """
        c = [i * j for i, j in zip(a, b)]
        return c

    P1 = xyz2llh(X, Y, Z, a=6378137.0, finv=298.257223563, output_degrees=False)
    P2 = xyz2llh(x, y, z, a=6378137.0, finv=298.257223563, output_degrees=False)
    # Subtract the reference coordinate from the observed coordinate
    P_diff = sub(P2, P1)
    # Convert the dLLH to dNEU
    # Assumes a sphere
    rot = [6378137.0, 6378137.0*cos(P1[0]), 1]
    NEU = mul(P_diff, rot)

    return tuple(NEU)


def covrot(X, Y, Z, cxx, cyy, czz, cxy, cxz, cyz):
    """
    Rotates the sigmas and the correlations from xyz to enu
    :param cxx:
    :param cyy:
    :param czz:
    :param cxy:
    :param cxz:
    :param cyz:
    :return:
    """
    def matmult(a, b):
        zip_b = zip(*b)
        return [[sum(ele_a*ele_b for ele_a, ele_b in zip(row_a, col_b)) for col_b in zip_b] for row_a in a]

    lat, lon, h = xyz2llh(X, Y, Z, a=6378137.0, finv=298.257223563, output_degrees=False)

    rxx = -sin(lat)*cos(lon)
    rxy = -sin(lat)*sin(lon)
    rxz = cos(lat)
    ryx = -sin(lon)
    ryy = cos(lon)
    ryz = 0
    rzx = cos(lat)*cos(lon)
    rzy = cos(lat)*sin(lon)
    rzz = sin(lat)

    rotation_matrix = [[rxx, rxy, rxz],
                       [ryx, ryy, ryz],
                       [rzx, rzy, rzz]]
    rotation_transpose = [[rxx, ryx, rzx],
                          [rxy, ryy, rzy],
                          [rxz, ryz, rzz]]
    covariance_matrix = [[cxx**2, cxy*cxx*cyy, cxz*cxx*czz],
                         [cxy*cxx*cyy, cyy**2, cyz*cyy*czz],
                         [cxz*cxx*czz, cyz*cyy*czz, czz**2]]

    tmp = matmult(rotation_matrix, covariance_matrix)

    result = matmult(tmp, rotation_transpose)

    cxx = result[0][0]**0.5
    cyy = result[1][1]**0.5
    czz = result[2][2]**0.5
    cxy = result[0][1]/(cxx*cyy)
    cxz = result[0][2]/(cxx*czz)
    cyz = result[1][2]/(cyy*czz)

    return cxx, cyy, czz, cxy, cxz, cyz

scripts/gsof_driver.py
#!/usr/bin/env python
"""
This script parses and converts Trimble GSOF messages incoming from a receiver and publishes the relevant ROS messages.
It  has been adapted from https://kb.unavco.org/kb/article/trimble-netr9-receiver-gsof-messages-806.html

@author: Michael Hutchinson (mhutchinson@sagarobotics.com)
"""


from struct import unpack
from trimble_gnss_driver.parser import parse_maps
from trimble_gnss_driver.gps_qualities import gps_qualities
import socket
import sys
import math
import rospy

from geometry_msgs.msg import Quaternion
from sensor_msgs.msg import NavSatFix, NavSatStatus, Imu # For lat lon h
from tf.transformations import quaternion_from_euler
import tf

"""
GSOF messages from https://www.trimble.com/OEM_ReceiverHelp/#GSOFmessages_Overview.html?TocPath=Output%2520Messages%257CGSOF%2520Messages%257COverview%257C_____0
"""
# Records we most care about
ATTITUDE = 27              # Attitude information with errors
LAT_LON_H = 2              # Position lat lon h
POSITION_SIGMA = 12        # Errors in position
POSITION_TYPE = 38
BASE_POSITION_QUALITY = 41 # Needed for gps quality indicator
INS_FULL_NAV = 49          # INS fused full nav info pose, attittude etc
INS_RMS = 50               # RMS errors from reported fused position
RECEIVED_BASE_INFO = 35    # Received base information

# Others (but still not the entire list of GSOF msgs available.)
VELOCITY = 8
SERIAL_NUM = 15
GPS_TIME = 1
UTC_TIME = 16
ECEF_POS = 3
LOCAL_DATUM = 4
LOCAL_ENU= 5

class GSOFDriver(object):
    """ A class to parse GSOF messages from a TCP stream. """

    def __init__(self):

        rospy.init_node('trimble_gnss_driver')

        port = rospy.get_param('~rtk_port', 21098)
        ip = rospy.get_param('~rtk_ip','192.168.0.50')

        self.output_frame_id = rospy.get_param("~output_frame_id", "base_link")

        apply_dual_antenna_offset = rospy.get_param("~apply_dual_antenna_offset", False)

        if apply_dual_antenna_offset:
            gps_main_frame_id = rospy.get_param('~gps_main_frame_id', 'gps_link')
            gps_aux_frame_id = rospy.get_param('~gps_aux_frame_id', 'gps_link')
            self.heading_offset = self.get_heading_offset(gps_main_frame_id, gps_aux_frame_id)
        else:
            self.heading_offset = 0.0
        rospy.loginfo("Heading offset is %f", self.heading_offset)

        self.fix_pub = rospy.Publisher('fix', NavSatFix, queue_size=1)
        # For attitude, use IMU msg to keep compatible with robot_localization
        # But note that this is not only from an IMU
        self.attitude_pub = rospy.Publisher('attitude', Imu, queue_size=1)
        # yaw from the dual antennas fills an Imu msg simply to keep consistent
        # with the current setup
        # Keep separate from attitude to avoid accidentally fusing zeros when
        # we don't measure roll/pitch
        self.yaw_pub = rospy.Publisher('yaw', Imu, queue_size=1)

        self.client = self.setup_connection(ip , port)

        self.msg_dict = {}
        self.msg_bytes = None
        self.checksum = True
        self.rec_dict = {}

        self.current_time = rospy.get_time()
        self.ins_rms_ts = 0
        self.pos_sigma_ts = 0
        self.quality_ts = 0
        self.error_info_timeout = 1.0
        self.base_info_timeout = 5.0


        while not rospy.is_shutdown():
            # READ GSOF STREAM
            self.records = []
            self.current_time = rospy.get_time()
            self.get_message_header()
            self.get_records()

            if not self.checksum:
                rospy.logwarn("Invalid checksum, skipping")
                continue

            # Make sure we have the information required to publish msgs and
            # that its not too old
            if INS_FULL_NAV in self.records:
                if INS_RMS in self.records or self.current_time - self.ins_rms_ts < self.error_info_timeout:
                    # print "Full INS info, filling ROS messages"
                    self.send_ins_fix()
                    self.send_ins_attitude()
                else:
                    rospy.logwarn("Skipping INS output as no matching errors within the timeout. Current time: %f, last error msg %f", self.current_time, self.ins_rms_ts)
            else:
                if LAT_LON_H in self.records:
                    if (POSITION_SIGMA in self.records or self.current_time - self.pos_sigma_ts < self.error_info_timeout) and (POSITION_TYPE in self.records or self.current_time - self.quality_ts < self.base_info_timeout):
                        self.send_fix()
                    else:
                        rospy.logwarn("Skipping fix output as no corresponding sigma errors or gps quality within the timeout. Current time: %f, last sigma msg %f, last gps quality msg %f", self.current_time, self.pos_sigma_ts, self.quality_ts)
                if ATTITUDE in self.records:
                    self.send_yaw()
            
            # if RECEIVED_BASE_INFO in self.records or LOCAL_DATUM in self.records or LOCAL_ENU in self.records:
            #     print(self.rec_dict)

                # print("Base Info: \n", self.rec_dict['BASE_NAME_1'], self.rec_dict['BASE_ID_1'], self.rec_dict['BASE_LATITUDE'], self.rec_dict['BASE_LONGITUDE'], self.rec_dict['BASE_HEIGHT'])
            # if INS_FULL_NAV in self.records and LAT_LON_H in self.records:
            #     print("Altitude INS: ", self.rec_dict['FUSED_ALTITUDE'], "Height LLH (WGS84): ", self.rec_dict['HEIGHT_WGS84'])


    def send_ins_fix(self):
        if self.rec_dict['FUSED_LATITUDE'] == 0 and self.rec_dict['FUSED_LONGITUDE']  == 0 and self.rec_dict['FUSED_ALTITUDE'] == 0:
            rospy.logwarn("Invalid fix, skipping")
            return
        current_time = rospy.get_rostime() # Replace with GPS time?
        fix = NavSatFix()

        fix.header.stamp = current_time
        fix.header.frame_id = self.output_frame_id

        gps_qual = gps_qualities[self.rec_dict['GPS_QUALITY']]
        fix.status.service = NavSatStatus.SERVICE_GPS # TODO: Fill correctly
        fix.status.status = gps_qual[0]
        fix.position_covariance_type = gps_qual[1]

        fix.latitude = self.rec_dict['FUSED_LATITUDE']
        fix.longitude = self.rec_dict['FUSED_LONGITUDE']

        # To follow convention set in the NavSatFix definition, altitude should be:
        # Altitude [m]. Positive is above the WGS 84 ellipsoid
        # Ref - http://docs.ros.org/en/api/sensor_msgs/html/msg/NavSatFix.html

        fix.altitude = self.rec_dict['FUSED_ALTITUDE']  # <-- CHECK

        fix.position_covariance[0] = self.rec_dict['FUSED_RMS_LONGITUDE'] ** 2
        fix.position_covariance[4] = self.rec_dict['FUSED_RMS_LATITUDE'] ** 2
        fix.position_covariance[8] = self.rec_dict['FUSED_RMS_ALTITUDE'] ** 2

        self.fix_pub.publish(fix)


    def send_ins_attitude(self):
        """
        We send the GNSS fused attitude information as an IMU msg to
        keep compatible with the robot_localization package. Although it should
        be noted that this is not just from an IMU but fused with dual gnss
        antennas information.
        """

        if self.rec_dict['FUSED_ROLL'] == 0 and self.rec_dict['FUSED_PITCH']  == 0 and self.rec_dict['FUSED_YAW'] == 0:
            rospy.logwarn("Invalid yaw, skipping")
            return

        current_time = rospy.get_rostime() # Replace with GPS time?

        # print "current time: ", current_time, "INS time: ", self.ins_rms_ts
        attitude = Imu()

        attitude.header.stamp = current_time
        attitude.header.frame_id = self.output_frame_id  # Assume transformation handled by receiver

        heading_enu = 2*math.pi - self.normalize_angle(math.radians(self.rec_dict['FUSED_YAW']) + 3*math.pi/2)
        orientation_quat = quaternion_from_euler(math.radians(self.rec_dict['FUSED_ROLL']),     #  roll sign stays the same
                                             - math.radians(self.rec_dict['FUSED_PITCH']),  # -ve for robots coord system (+ve down)
                                             heading_enu)
        # print 'r p y_enu receiver_heading [degs]: ', self.rec_dict['FUSED_ROLL'], self.rec_dict['FUSED_PITCH'], math.degrees(heading_enu), self.rec_dict['FUSED_YAW']

        attitude.orientation = Quaternion(*orientation_quat)

        attitude.orientation_covariance[0] = math.radians(self.rec_dict['FUSED_RMS_ROLL']) ** 2  # [36] size array
        attitude.orientation_covariance[4] = math.radians(self.rec_dict['FUSED_RMS_PITCH']) ** 2  # [36] size array
        attitude.orientation_covariance[8] = math.radians(self.rec_dict['FUSED_RMS_YAW']) ** 2 # [36] size array

        self.attitude_pub.publish(attitude)


    def send_fix(self):
        if self.rec_dict['LATITUDE'] == 0 and self.rec_dict['LONGITUDE']  == 0 and self.rec_dict['HEIGHT_WGS84'] == 0:
            rospy.logwarn("Invalid fix, skipping")
            return

        current_time = rospy.get_rostime() # Replace with GPS time?
        fix = NavSatFix()

        fix.header.stamp = current_time
        fix.header.frame_id = self.output_frame_id
        gps_qual = self.get_gps_quality()
        fix.status.service = NavSatStatus.SERVICE_GPS # TODO: Fill correctly
        fix.status.status = gps_qual[0]
        fix.position_covariance_type = gps_qual[1]

        fix.latitude = math.degrees(self.rec_dict['LATITUDE'])
        fix.longitude = math.degrees(self.rec_dict['LONGITUDE'])

        # To follow convention set in the NavSatFix definition, altitude should be:
        # Altitude [m]. Positive is above the WGS 84 ellipsoid
        # Ref - http://docs.ros.org/en/api/sensor_msgs/html/msg/NavSatFix.html

        fix.altitude = self.rec_dict['HEIGHT_WGS84']  # <-- CHECK

        fix.position_covariance[0] = self.rec_dict['SIG_EAST'] ** 2  # Check east north order
        fix.position_covariance[4] = self.rec_dict['SIG_NORT'] ** 2
        fix.position_covariance[8] = self.rec_dict['SIG_UP'] ** 2

        self.fix_pub.publish(fix)


    def send_yaw(self):
        """
        We send yaw (without ins) as an IMU message for compatibility with our
        other software
        """
        if self.rec_dict['ROLL'] == 0 and self.rec_dict['PITCH']  == 0 and self.rec_dict['YAW'] == 0:
            rospy.logwarn("Invalid yaw, skipping")
            return

        current_time = rospy.get_rostime() # Replace with GPS time?

        # print "current time: ", current_time, "INS time: ", self.ins_rms_ts
        yaw = Imu()

        yaw.header.stamp = current_time
        yaw.header.frame_id = self.output_frame_id  # Assume transformation handled by receiver

        heading_ned = self.normalize_angle(self.rec_dict['YAW'] + self.heading_offset)
        heading_enu = 2*math.pi - self.normalize_angle(heading_ned + 3*math.pi/2)

        orientation_quat = quaternion_from_euler(self.rec_dict['ROLL'],     #  roll sign stays the same
                                             - self.rec_dict['PITCH'],  # -ve for robots coord system (+ve down)
                                             heading_enu)
        # print 'r p y receiver_heading [rads]: ', self.rec_dict['ROLL'], self.rec_dict['PITCH'], heading_enu, self.rec_dict['YAW']

        yaw.orientation = Quaternion(*orientation_quat)

        yaw.orientation_covariance[0] = self.rec_dict['ROLL_VAR'] ** 2  # [36] size array
        yaw.orientation_covariance[4] = self.rec_dict['PITCH_VAR'] ** 2  # [36] size array
        yaw.orientation_covariance[8] = self.rec_dict['YAW_VAR'] ** 2 # [36] size array

        self.yaw_pub.publish(yaw)


    def get_gps_quality(self):
        """Get ROS NavSatStatus position type from trimbles
        """
        trimble_position_type = self.rec_dict['POSITION_TYPE']

        if trimble_position_type >= 9:
            position_type = 4 # fix
        elif trimble_position_type >= 7:
            position_type = 5 # float
        elif trimble_position_type >= 1:
            position_type = 2
        else:
            position_type = 0

        return gps_qualities[position_type]


    @staticmethod
    def normalize_angle(angle_in):
        while angle_in > 2*math.pi:
            angle_in = angle_in - 2*math.pi
        while angle_in < 0:
            angle_in = angle_in + 2*math.pi
        return angle_in


    def get_heading_offset(self, gps_main_frame_id, gps_aux_frame_id):

        if gps_main_frame_id == gps_aux_frame_id:
            rospy.logerr("Cannot offset antenna yaw if they have the same frame_id's, will assume 0.0")
            return 0.0

        #        Get GPS antenna tf
        tf_listener = tf.TransformListener()

        rospy.loginfo( "Waiting for GPS tf between antennas.")
        got_gps_tf = False
        while not got_gps_tf:
            try:
                (trans, rot) = tf_listener.lookupTransform(gps_main_frame_id, gps_aux_frame_id, rospy.Time(0))
                got_gps_tf = True
            except (tf.LookupException, tf.ConnectivityException, tf.ExtrapolationException):
                got_gps_tf = False
                continue

        rospy.loginfo( "Got gps tf")
        dy_antennas = trans[1]
        dx_antennas = trans[0]

        if got_gps_tf:
            heading_offset = math.atan2(dy_antennas,dx_antennas)
        else:
            heading_offset = 0

        return heading_offset


    def setup_connection(self, _ip, _port):
        port = _port
        ip = None
        attempts_limit = 10
        current_attempt = 0
        connected = False

        client = socket.socket(socket.AF_INET, socket.SOCK_STREAM)

        while not connected and not rospy.is_shutdown() and current_attempt < attempts_limit:
            current_attempt += 1

            try:
                client = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
                client.settimeout(5)
                ip = socket.gethostbyname(_ip)
                address = (ip, port)

                rospy.loginfo("Attempting connection to %s:%s ", ip, port)
                client.connect(address)

                rospy.loginfo("=====================================")
                rospy.loginfo("Connected to %s:%s ", ip, port)
                rospy.loginfo("=====================================")
                connected = True
            except Exception as e:
                rospy.logwarn("Connection to IP: " + ip + ": " + e.__str__() +
                              ".\nRetrying connection: Attempt: %s/%s",
                              current_attempt, attempts_limit)

        if not connected:
            rospy.logerr("No connection established. Node shutting down")
            sys.exit()

        return client


    def get_message_header(self):
        data = self.client.recv(7)
        msg_field_names = ('STX', 'STATUS', 'TYPE', 'LENGTH',
                           'T_NUM', 'PAGE_INDEX', 'MAX_PAGE_INDEX')
        self.msg_dict = dict(zip(msg_field_names, unpack('>7B', data)))
        # print "msg dict: ", self.msg_dict
        self.msg_bytes = self.client.recv(self.msg_dict['LENGTH'] - 3)
        (checksum, etx) = unpack('>2B', self.client.recv(2))

        if checksum-self.checksum256(self.msg_bytes+data[1:]) == 0:
            self.checksum = True
        else:
            self.checksum = False
        # print("Checksum: ", self.checksum)


    def checksum256(self, st):
        """Calculate checksum"""
        if sys.version_info[0] >= 3:
            return sum(st) % 256
        else:
            return reduce(lambda x, y: x+y, map(ord, st)) % 256


    def get_records(self):
        self.byte_position = 0
        while self.byte_position < len(self.msg_bytes):
            # READ THE FIRST TWO BYTES FROM RECORD HEADER
            record_type, record_length = unpack('>2B', self.msg_bytes[self.byte_position:self.byte_position + 2])
            self.byte_position += 2
            self.records.append(record_type)
            # print "Record type: ", record_type, " Length: ", record_length
            # self.select_record(record_type, record_length)
            if record_type in parse_maps:
                self.rec_dict.update(dict(zip(parse_maps[record_type][0], unpack(parse_maps[record_type][1], self.msg_bytes[self.byte_position:self.byte_position + record_length]))))
            else:
                rospy.logwarn("Record type %s is not in parse maps.", record_type)
            self.byte_position += record_length

        if INS_RMS in self.records:
            self.ins_rms_ts = self.current_time
        if POSITION_SIGMA in self.records:
            self.pos_sigma_ts = self.current_time
        if POSITION_TYPE in self.records:
            self.quality_ts = self.current_time

if __name__ == '__main__':
    try:
        GSOFDriver()
    except rospy.ROSInterruptException:
        pass
